pkg/sensors/exec/exec_test.go

```diff
@@ -945,23 +948,125 @@ func TestExecProcessCredentialsSuid(t *testing.T) {
+	 * so no privilege change should be detected, same filters as above apply.
+	 */
+	if err := os.Chown(testSuid, 0, 0); err != nil {
+		t.Fatalf("Chown() on '%s' binary error: %s\n", testSuid, err)
+	}
+	if err := os.Chmod(testSuid, 0754|os.ModeSetuid|os.ModeSetgid); err != nil {
+		t.Fatalf("Chown() on '%s' binary error: %s\n", testSuid, err)
+	}
+	procExecSetuidRootNoPrivsChanged := ec.NewProcessChecker().
+		WithBinary(sm.Full(testSuid)).WithProcessCredentials(noCredsChange).WithBinaryProperties(nil)
+	execSetuidRootNoPrivsChangedChecker := ec.NewProcessExecChecker("exec").WithProcess(procExecSetuidRootNoPrivsChanged)
+	if err := exec.Command(testSuid).Run(); err != nil {
+		t.Fatalf("Failed to execute '%s' binary: %s\n", testSuid, err)
+	}
+
+	checker := ec.NewUnorderedEventChecker(execNoPrivsChangedChecker, execSetuidRootNoPrivsChangedChecker)
+	err = jsonchecker.JsonTestCheck(t, checker)
+	assert.NoError(t, err)
+}
+
+// Test running with different combinations of setgid bit set
+//  1. setgid() systemcall to arbitrary gid value then exec binary to
+//     assert credentials did not change.
+//  2. executes a set-group-ID to root binary asserting that we detect
+//     the setgid bit set + the privileges changed due to the setgid bit
+//     being set to root group.
+//  3. executes a set-group-ID to non root, it is set to arbitrary value
+//     to assert that binary execution detects the setgid bit but we do
+//     not report as a privilege changed execution as the target group
+//     is not root.
+func TestExecProcessCredentialsSetgidChanges(t *testing.T) {
+	var doneWG, readyWG sync.WaitGroup
+	defer doneWG.Wait()
+
+	ctx, cancel := context.WithTimeout(context.Background(), tus.Conf().CmdWaitTime)
+	defer cancel()
+
+	obs, err := observertesthelper.GetDefaultObserver(t, ctx, tus.Conf().TetragonLib, observertesthelper.WithMyPid())
+	if err != nil {
+		t.Fatalf("Failed to run observer: %s", err)
+	}
+
+	testBin := testutils.RepoRootPath("contrib/tester-progs/nop")
+	// We should be able to create suid on local mount point
+	testSuid := testutils.RepoRootPath("contrib/tester-progs/suidnop")
+	if err := testutils.CopyFile(testSuid, testBin, 0754|os.ModeSetuid|os.ModeSetgid); err != nil {
+		t.Fatalf("Failed to copy binary: %s", err)
+	}
+
+	t.Cleanup(func() {
+		if err = syscall.Setuid(0); err != nil {
+			t.Fatalf("Failed to restore uid to 0 :  %s\n", err)
+		}
+		if err = syscall.Setgid(0); err != nil {
+			t.Fatalf("Failed to restore gid to 0 :  %s\n", err)
+		}
+
+		err := os.Remove(testSuid)
+		if err != nil {
+			t.Logf("Error failed to cleanup '%s'", testSuid)
+		}
+	})
 
+	observertesthelper.LoopEvents(ctx, t, &doneWG, &readyWG, obs)
+	readyWG.Wait()
+
+	/* Executing a setgid to root with current gid as normal not root */
 	gid := 1879048188
 	if err := syscall.Setgid(gid); err != nil {
-		t.Fatalf("setegid(%d) error: %s", gid, err)
+		t.Fatalf("setgid(%d) error: %s", gid, err)
 	}
 
```

Comment:
Adding my previous  [comment](https://github.com/cilium/tetragon/pull/1786#discussion_r1431248693) for context:
> I'm not seeing any Setuid() calls or Setgid calls in this test. There are some in the test below, so maybe that's what it tries to fix? In any case, I think we should add a t.Cleanup() for resetting the set{u,g}ids wherever we change them so that t is clear (i) why the happen, (ii) that the set values will be restored.

What I meant was that the t.Cleanup() should be added wherever we change the Setuid.

For example:

```go
oldGid := syscall.Getgid()
if err := syscall.Setgid(gid); err != nil {
  ...
}
t.Cleanup(func() {
    		if err = syscall.Setgid(oldGid); err != nil {
			t.Logf("Failed to restore gid to %d :  %v\n", oldGid, err)
		}
}) 
```

Above makes it clear what the cleanup functions fixes.

We can even have helper functions for this:

```go
func TestSetGID(t *testing.T, gid int)  {

    oldGid := syscall.Getgid()
   if err := syscall.Setgid(gid); err != nil {
      t.Fail("failed to set gid to %d", gid) // can also be t.Skip if we want to
   }

   t.Cleanup(func() {
    		if err = syscall.Setgid(oldGid); err != nil {
			t.Logf("Failed to restore gid to %d :  %v\n", oldGid, err)
		}
   }) 
}
```

```